File: src/octop/api/routers/workbuddy_lifecycle.py
"""WorkBuddy tenant lifecycle API: controlled export, one-time redeem, deletion.

Paths are relative to the ``/api/v1`` mount (see ``contracts/route-manifest.json``):

* ``POST /tenants/{id}/export`` — 202, builds the redacted export and mints the
  single redeem token for the fixed 72h window;
* ``POST /exports/{id}/redeem`` — 200, one-time CAS consumption of that token;
* ``GET  /exports/{id}/manifest`` — the frozen manifest and its digest;
* ``POST /tenants/{id}/deletion-requests`` — 201 inside cooling-off, or
  ``COMPLIANCE_GATE_CLOSED`` when no signed policy authorises the tenant;
* ``GET  /tenants/{id}/deletion-requests/{request_id}`` — stage and timestamps;
* ``POST /tenants/{id}/deletion-requests/{request_id}/cancel`` — CAS cancel,
  only while the cooling-off window is open.

Tenant identity always comes from the authenticated principal: the ``{id}`` in
the path must equal the principal's tenant and anything else is reported as a
uniform 404.  Deletion requests additionally require a recent login (``iat``
within :data:`RECENT_AUTH_MAX_AGE_SECONDS`, otherwise 428) and never queue work
when the compliance gate is closed.  Export payloads and redeem tokens are
returned ``no-store`` and are never logged.
"""

from __future__ import annotations

import time
import uuid
from typing import Annotated, Any

from fastapi import APIRouter, Body, Depends, Request, Response
from pydantic import BaseModel, Field

from octop.api.deps import InvalidToken, TokenExpired, decode_claims, extract_raw_token, get_server
from octop.api.routers.workbuddy_identity import (
    WorkBuddyPrincipal,
    require_workbuddy_admin,
    workbuddy_envelope,
    workbuddy_principal,
)
from octop.infra.db.workbuddy_context import WorkBuddyPostgresRequiredError
from octop.infra.errors import ErrorCode, OctopError
from octop.infra.workbuddy import lifecycle as policy

router = APIRouter()

_Principal = Annotated[WorkBuddyPrincipal, Depends(workbuddy_principal)]
_AdminPrincipal = Annotated[WorkBuddyPrincipal, Depends(require_workbuddy_admin())]

RECENT_AUTH_MAX_AGE_SECONDS = 3600

_EXPORT_NOT_FOUND = "export job not found"
_DELETION_NOT_FOUND = "deletion request not found"
_NO_STORE = "no-store"


class RedeemBody(BaseModel):
    token: str = Field(min_length=8, max_length=512)


class DeletionCancelBody(BaseModel):
    expected_version: int | None = Field(default=None, ge=1)


def _repo(server: Any) -> Any:
    from octop.infra.db.repos.workbuddy_lifecycle import WorkBuddyLifecycleRepo

    return WorkBuddyLifecycleRepo(server.services.db)


def _not_found(detail: str) -> OctopError:
    return OctopError(ErrorCode.NOT_FOUND, detail)


def _public_id(value: str, detail: str) -> str:
    """Normalize a public id; a malformed id looks exactly like an absent one."""
    try:
        return str(uuid.UUID(str(value)))
    except (AttributeError, TypeError, ValueError) as exc:
        raise _not_found(detail) from exc


def _own_tenant(path_tenant_id: str, principal: WorkBuddyPrincipal) -> str:
    """A tenant admin only ever reaches their own tenant, and others get a 404."""
    tenant_id = _public_id(path_tenant_id, "tenant not found")
    if tenant_id != str(principal.tenant_id):
        raise _not_found("tenant not found")
    return tenant_id


def _require_recent_auth(request: Request, server: Any) -> None:
    """Deletion and cancel require a fresh login (428 when the token is old)."""
    raw = extract_raw_token(authorization=request.headers.get("authorization"))
    if not raw:
        raise OctopError(
            ErrorCode.PRECONDITION_REQUIRED, "recent authentication required for this action"
        )
    try:
        claims = decode_claims(server, raw)
    except (InvalidToken, TokenExpired) as exc:
        raise OctopError(
            ErrorCode.PRECONDITION_REQUIRED, "recent authentication required for this action"
        ) from exc
    issued_at = claims.get("iat")
    if not isinstance(issued_at, (int, float)):
        raise OctopError(
            ErrorCode.PRECONDITION_REQUIRED, "recent authentication required for this action"
        )
    if int(time.time()) - int(issued_at) > RECENT_AUTH_MAX_AGE_SECONDS:
        raise OctopError(
            ErrorCode.PRECONDITION_REQUIRED,
            "recent authentication required for this action; sign in again",
        )


def _no_store(response: Response) -> None:
    response.headers["Cache-Control"] = _NO_STORE
    response.headers["Pragma"] = _NO_STORE


def _translate(exc: WorkBuddyPostgresRequiredError) -> OctopError:
    return OctopError(ErrorCode.WORKBUDDY_POSTGRES_REQUIRED, str(exc))


# ── export ──────────────────────────────────────────────────────────────────


@router.post("/tenants/{tenant_id}/export", status_code=202, summary="Request a tenant data export")
async def request_tenant_export(
    request: Request,
    tenant_id: str,
    response: Response,
    principal: _AdminPrincipal,
    server: Any = Depends(get_server),
) -> dict[str, Any]:
    """Build the redacted export and return the job id plus its one redeem token."""
    tenant = _own_tenant(tenant_id, principal)
    try:
        issue = policy.start_tenant_export(
            _repo(server), tenant_id=tenant, user_id=principal.user_id
        )
    except WorkBuddyPostgresRequiredError as exc:
        raise _translate(exc) from exc
    _no_store(response)
    return workbuddy_envelope(request, issue.as_payload())


@router.get("/exports/{export_job_id}/manifest", summary="Read the frozen export manifest")
async def read_export_manifest(
    request: Request,
    export_job_id: str,
    response: Response,
    principal: _AdminPrincipal,
    server: Any = Depends(get_server),
) -> dict[str, Any]:
    job_id = _public_id(export_job_id, _EXPORT_NOT_FOUND)
    try:
        manifest = policy.read_export_manifest(
            _repo(server), tenant_id=str(principal.tenant_id), export_job_id=job_id
        )
    except WorkBuddyPostgresRequiredError as exc:
        raise _translate(exc) from exc
    if manifest is None:
        raise _not_found(_EXPORT_NOT_FOUND)
    _no_store(response)
    return workbuddy_envelope(request, manifest)


@router.post("/exports/{export_job_id}/redeem", summary="Redeem an export token once")
async def redeem_export(
    request: Request,
    export_job_id: str,
    response: Response,
    body: RedeemBody,
    principal: _AdminPrincipal,
    server: Any = Depends(get_server),
) -> dict[str, Any]:
    """Consume the one-time token; a second attempt is refused, never replayed."""
    job_id = _public_id(export_job_id, _EXPORT_NOT_FOUND)
    try:
        download = policy.redeem_export(
            _repo(server),
            tenant_id=str(principal.tenant_id),
            user_id=principal.user_id,
            token=body.token,
        )
    except WorkBuddyPostgresRequiredError as exc:
        raise _translate(exc) from exc
    if download.export_job_id != job_id:
        # The token belongs to another job of this tenant: report it as not found.
        raise _not_found(_EXPORT_NOT_FOUND)
    _no_store(response)
    return workbuddy_envelope(
        request,
        {
            "export_job_id": download.export_job_id,
            "manifest": download.manifest,
            "manifest_sha256": download.manifest_sha256,
            "tables": list(download.tables),
            "redeemed_at": download.redeemed_at,
        },
    )


# ── deletion ────────────────────────────────────────────────────────────────


@router.post(
    "/tenants/{tenant_id}/deletion-requests",
    status_code=201,
    summary="Request tenant deletion (30-day cooling-off)",
)
async def create_deletion_request(
    request: Request,
    tenant_id: str,
    principal: _AdminPrincipal,
    server: Any = Depends(get_server),
) -> dict[str, Any]:
    """Open the cooling-off window; without a signed policy this answers 503."""
    tenant = _own_tenant(tenant_id, principal)
    _require_recent_auth(request, server)
    try:
        view = policy.request_tenant_deletion(
            _repo(server), tenant_id=tenant, user_id=principal.user_id
        )
    except WorkBuddyPostgresRequiredError as exc:
        raise _translate(exc) from exc
    return workbuddy_envelope(request, view.as_payload())


@router.get(
    "/tenants/{tenant_id}/deletion-requests/{deletion_request_id}",
    summary="Read a deletion request stage and timestamps",
)
async def read_deletion_request(
    request: Request,
    tenant_id: str,
    deletion_request_id: str,
    principal: _AdminPrincipal,
    server: Any = Depends(get_server),
) -> dict[str, Any]:
    tenant = _own_tenant(tenant_id, principal)
    request_id = _public_id(deletion_request_id, _DELETION_NOT_FOUND)
    try:
        view = policy.read_deletion_request(
            _repo(server), tenant_id=tenant, deletion_request_id=request_id
        )
    except WorkBuddyPostgresRequiredError as exc:
        raise _translate(exc) from exc
    return workbuddy_envelope(request, view.as_payload())


@router.post(
    "/tenants/{tenant_id}/deletion-requests/{deletion_request_id}/cancel",
    summary="Cancel a deletion request inside its cooling-off window",
)
async def cancel_deletion_request(
    request: Request,
    tenant_id: str,
    deletion_request_id: str,
    principal: _AdminPrincipal,
    server: Any = Depends(get_server),
    body: Annotated[DeletionCancelBody | None, Body()] = None,
) -> dict[str, Any]:
    """CAS cancel: only a live cooling-off request with a matching version wins."""
    tenant = _own_tenant(tenant_id, principal)
    request_id = _public_id(deletion_request_id, _DELETION_NOT_FOUND)
    _require_recent_auth(request, server)
    expected_version = None if body is None else body.expected_version
    try:
        view = policy.cancel_tenant_deletion(
            _repo(server),
            tenant_id=tenant,
            user_id=principal.user_id,
            deletion_request_id=request_id,
            expected_version=expected_version,
        )
    except WorkBuddyPostgresRequiredError as exc:
        raise _translate(exc) from exc
    return workbuddy_envelope(request, view.as_payload())

File: dashboard/src/api/modules/workbuddyLifecycle.ts
/**
 * WorkBuddy tenant lifecycle API — controlled export, one-time redeem, deletion.
 * All paths are relative to ``/api/v1`` (``request()`` adds the ``/api`` prefix).
 *
 * Frozen manifest (contracts/route-manifest.json, stages C/D):
 *   POST /v1/tenants/{id}/export
 *   POST /v1/exports/{id}/redeem
 *   POST /v1/exports/{id}/download-challenge              (stage D)
 *   POST /v1/auth/reauthenticate                          (stage D)
 *   POST /v1/tenants/{id}/deletion-requests
 *   GET  /v1/tenants/{id}/deletion-requests/{request_id}
 *   POST /v1/tenants/{id}/deletion-requests/{request_id}/cancel
 *
 * Field names are taken verbatim from the slice on
 * ``feat/workbuddy-compliance-lifecycle``: the router
 * (``octop.api.routers.workbuddy_lifecycle``), ``ExportIssue.as_payload`` /
 * ``ExportDownload`` / ``DeletionView.as_payload`` in
 * ``octop.infra.workbuddy.lifecycle`` and the job status vocabulary in
 * ``octop.infra.db.repos.workbuddy_lifecycle``. The two stage-D routes have no
 * implementation on any branch yet, so their request/response shapes are
 * derived from the manifest rows only and are marked "provisional" below.
 *
 * Two invariants this module never breaks:
 *  - The redeem token returned by ``requestTenantExport`` is the only raw token
 *    this module ever produces; it is handed back to the caller and never
 *    stored, echoed into a path, or logged here.
 *  - Deletion is fail-closed: ``COMPLIANCE_GATE_CLOSED`` is surfaced as an
 *    error code so the UI can show that the entry point is closed. Nothing in
 *    this module synthesises a queued or optimistic deletion state.
 */

import { request } from "../request";

const BASE = "/v1";

/** Every success body is wrapped: ``{ data, request_id }``. */
export interface ApiEnvelope<T> {
  data: T;
  request_id?: string;
}

async function unwrap<T>(path: string, init?: RequestInit): Promise<T> {
  const body = await request<ApiEnvelope<T>>(path, init);
  return body.data;
}

function jsonInit(method: string, body?: unknown): RequestInit {
  return body === undefined
    ? { method }
    : { method, body: JSON.stringify(body) };
}

/** ``ErrorCode`` members this surface answers with (see ``octop.infra.errors``). */
export const COMPLIANCE_GATE_CLOSED = "COMPLIANCE_GATE_CLOSED";
export const DELETION_CANCEL_WINDOW_CLOSED = "DELETION_CANCEL_WINDOW_CLOSED";
export const DELETION_REQUEST_CONFLICT = "DELETION_REQUEST_CONFLICT";
export const PRECONDITION_REQUIRED = "PRECONDITION_REQUIRED";

/**
 * True when the server has no such surface yet: an unmounted slice answers 404
 * and an unimplemented feature 501. 503 is deliberately *not* treated as
 * "unavailable" — this slice answers ``COMPLIANCE_GATE_CLOSED`` with 503, and
 * that must stay visible as a gate state, never as an empty panel. A transport
 * failure is not "unavailable" either: it falls through to the normal error
 * path instead of claiming the feature does not exist.
 */
export function isLifecycleUnavailableError(error: unknown): boolean {
  const text = error instanceof Error ? error.message : String(error ?? "");
  return (
    /\b404\b/.test(text) ||
    /\b501\b/.test(text) ||
    /not implemented/i.test(text)
  );
}

// --- Export -----------------------------------------------------------------

/** Job status vocabulary of the lifecycle repo (``queued → ready → …``). */
export type ExportJobStatus =
  | "queued"
  | "ready"
  | "failed"
  | "redeemed"
  | "expired";

/**
 * ``ExportIssue.as_payload()`` — creation result of ``POST /tenants/{id}/export``
 * (202). ``redeem_token`` is the single raw one-time token; the server never
 * returns it again and keeps only its sha256.
 */
export interface TenantExportIssue {
  job_id: string;
  export_job_id: string;
  status: ExportJobStatus;
  redeem_token: string;
  redeem_expires_at: number;
  manifest_sha256: string | null;
  table_total: number;
  row_total: number;
}

/** One exported table entry of the frozen manifest (``ExportTable``). */
export interface ExportManifestTable {
  name: string;
  row_count: number;
  columns: string[];
  excluded_columns: string[];
  content_sha256: string;
}

/** One table kept out of the export entirely (``ExportTableExclusion``). */
export interface ExportManifestExclusion {
  name: string;
  category: string;
  included: false;
}

/** Canonical manifest whose sha256 is recorded with the job. */
export interface ExportManifest {
  schema: string;
  redaction_rules_version: number;
  tenant_id: string;
  export_job_id: string;
  created_at: number;
  tables: ExportManifestTable[];
  excluded_tables: ExportManifestExclusion[];
  totals: { tables: number; rows: number; excluded_tables: number };
  content_sha256: string;
}

/** One table of a redeemed payload; rows are already redacted server-side. */
export interface ExportRedeemedTable {
  name: string;
  row_count: number;
  content_sha256: string;
  rows: Record<string, unknown>[];
}

/** Body of a consumed redeem token (``POST /exports/{id}/redeem``, 200). */
export interface ExportDownload {
  export_job_id: string;
  manifest: ExportManifest;
  manifest_sha256: string;
  tables: ExportRedeemedTable[];
  redeemed_at: number;
}

/**
 * Provisional: ``POST /auth/reauthenticate`` (stage D) is documented in the
 * manifest as "issues a five-minute one-time re-authentication credential"
 * bound to tenant/user/purpose. No router implements it on any branch yet, so
 * only the purpose of each field is frozen — treat a mismatch as a server
 * validation error, never as a client-side truth.
 */
export interface ReauthenticateRequest {
  password: string;
  purpose?: string;
}

/** Provisional — see ``ReauthenticateRequest``. */
export interface ReauthenticateResult {
  credential: string;
  expires_at: number;
}

/**
 * Provisional: ``POST /exports/{id}/download-challenge`` (stage D) issues the
 * export redeem challenge for a freshly re-authenticated admin, ``no-store``,
 * without extending the 72h redeem window.
 */
export interface ExportDownloadChallengeRequest {
  credential: string;
}

/** Provisional — see ``ExportDownloadChallengeRequest``. */
export interface ExportDownloadChallenge {
  challenge: string;
  expires_at: number;
}

// --- Deletion ---------------------------------------------------------------

/** Stage vocabulary of ``deletion_timeline`` in the lifecycle policy. */
export type DeletionStage = "cooling_off" | "cancelled" | "archived" | "purged";

/**
 * ``DeletionView.as_payload()`` — the only authority on a deletion request's
 * state. ``cancellable`` is server-computed: cancel is refused once the
 * cooling-off window has closed, a legal hold is active only when the server
 * says so, and nothing here derives either flag locally.
 */
export interface DeletionRequestView {
  deletion_request_id: string;
  tenant_id: string;
  stage: DeletionStage;
  requested_at: number;
  cooling_off_ends_at: number;
  purge_due_at: number | null;
  purged_at: number | null;
  cancellable: boolean;
  legal_hold_active: boolean;
  version: number;
  policy_sha256: string;
  policy_expires_at: number;
  archive_sha256: string | null;
  tombstone_sha256: string | null;
}

/** CAS guard for the cancel route; ``expected_version`` is the loaded view's. */
export interface DeletionCancelRequest {
  expected_version?: number;
}

export const workbuddyLifecycleApi = {
  /**
   * Build the redacted export and mint its single redeem token (202). The
   * caller must display the token exactly once and never persist it.
   */
  requestTenantExport: (tenantId: string) =>
    unwrap<TenantExportIssue>(
      `${BASE}/tenants/${encodeURIComponent(tenantId)}/export`,
      jsonInit("POST", {}),
    ),

  /** Consume the one-time token; a second attempt is refused by the server. */
  redeemExport: (exportJobId: string, token: string) =>
    unwrap<ExportDownload>(
      `${BASE}/exports/${encodeURIComponent(exportJobId)}/redeem`,
      jsonInit("POST", { token }),
    ),

  /** Provisional stage-D route; see ``ExportDownloadChallengeRequest``. */
  requestDownloadChallenge: (
    exportJobId: string,
    body: ExportDownloadChallengeRequest,
  ) =>
    unwrap<ExportDownloadChallenge>(
      `${BASE}/exports/${encodeURIComponent(exportJobId)}/download-challenge`,
      jsonInit("POST", body),
    ),

  /** Provisional stage-D route; see ``ReauthenticateRequest``. */
  reauthenticate: (body: ReauthenticateRequest) =>
    unwrap<ReauthenticateResult>(
      `${BASE}/auth/reauthenticate`,
      jsonInit("POST", body),
    ),

  /**
   * Open the 30-day cooling-off window (201). Without a signed compliance
   * policy the server answers ``COMPLIANCE_GATE_CLOSED`` and queues nothing.
   */
  createDeletionRequest: (tenantId: string) =>
    unwrap<DeletionRequestView>(
      `${BASE}/tenants/${encodeURIComponent(tenantId)}/deletion-requests`,
      jsonInit("POST", {}),
    ),

  /** Read a deletion request's stage and timestamps. */
  getDeletionRequest: (tenantId: string, deletionRequestId: string) =>
    unwrap<DeletionRequestView>(
      `${BASE}/tenants/${encodeURIComponent(
        tenantId,
      )}/deletion-requests/${encodeURIComponent(deletionRequestId)}`,
    ),

  /** CAS cancel: only a live cooling-off request with a matching version wins. */
  cancelDeletionRequest: (
    tenantId: string,
    deletionRequestId: string,
    body?: DeletionCancelRequest,
  ) =>
    unwrap<DeletionRequestView>(
      `${BASE}/tenants/${encodeURIComponent(
        tenantId,
      )}/deletion-requests/${encodeURIComponent(deletionRequestId)}/cancel`,
      jsonInit("POST", body ?? {}),
    ),
};
